syntax = "proto3";

package users;

import "general.proto";

/*
* Messages used to handle requests
* to the user endpoint
*/



// Message used to store the user data
// in the statefun and also what we give back to the
// queries
message UserData{
  string id = 1;
  int64 credit = 2;
}

// Message used to keep track of the current max
// id to be assigned
message Count{
  uint64 num = 1;
}

// USER CREATION MESSAGES

message CreateUserRequest{
  RequestInfo request_info = 1;
  string id = 2;
}





// Message containing the others
message UserRequest{

  // GET USER MESSAGE
  message FindUserRequest{
    string id = 1;
  }

  // USER DELETION MESSAGES
  message RemoveUserRequest{
    string id = 1;
  }

  // SUBTRACT CREDIT MESSAGES
  message SubtractCreditRequest{
    string id = 1;
    int64 amount = 2;
  }

  // ADD CREDIT MESSAGES
  message AddCreditRequest{
    string id = 1;
    int64 amount = 2;
  }

  RequestInfo request_info = 1;

  oneof message {
    FindUserRequest find_user = 3;
    RemoveUserRequest remove_user = 4;
    SubtractCreditRequest subtract_credit = 5;
    AddCreditRequest add_credit = 6;
  }

}


// INTERACTION WITH THE PAYMENT SERVICE

message UserPayRequest{
  int64 amount = 1;
  string order_id = 2;

  // Include the request info
  RequestInfo request_info = 3;
}

// Message to cancel the payment
message UserCancelPayRequest{
  int64 amount = 1;
  string order_id = 2;

  RequestInfo request_info = 3;
}

// Response returned to the payment
// service with success or failure
message UserPayResponse{
  bool success = 1;
  string order_id = 2;
  RequestInfo request_info = 3;
}